### PR TITLE
Github Tasks Rework

### PR DESCRIPTION
**Pull request by m-q-t**
Hi team,

Please find in this PR a re-work of the Github Tasks.

The following has been changed:
- `Gemfile` -> Add `octokit` dependency (official Github Ruby Client)

- `config/config.json.default` -> Remove `gitrob_access_token` and instead replace it with `github_access_token`
-  `data/gitleaks.config` -> The (default) signature file for `gitleaks` to use
- `lib/entities/github_repository.rb` -> Since this is specifically for Github, I prepended `https://github.com/*` so any discovered Github Repositories can be passed to the `gitleaks` task in a workflow. Maybe in the future we'll have an abstract `Git `repository entity where this can be changed.
- `lib/intrigue-tasks.rb` -> Require the new Github Client
-  `lib/tasks/base.rb` -> Include the new Github Helper Module
-  `lib/tasks/gather_github_repositories.rb` -> Uses the Github API to gather an account's Github Repositories. If a valid Github Access Token is provided this will use the authenticated API (thus allowing far more requests). If no `account` is provided in the options, this will retrieve all the public/private/organization repositories the token has access to. If an `account` is provided, this will retrieve all the public repositories for the account. Finally, if no valid access token is found this will use unauthenticated techniques to gather repositories for a specified account.
- `lib/tasks/gitleaks.rb ` -> The new task replacing `gitrob`. The way the `gitleaks` task works is that (any) URI which is passed in that contains a `.git` repository, `gitleaks` will be run against. The user has the ability to pass in custom keywords in which `gitleaks` will only match findings based on the respective keyword. 
- `lib/tasks/helpers/github.rb` -> The new Github helper module, this mainly contains helpers associated with `gitleaks` as its ran in two different tasks (more on that later).  There are also helpers which validate whether the access token is valid and if so, returns an initialized client.
- `lib/tasks/search_github_code.rb` -> This task uses authenticated techniques to search across all Github repositories for a unique keyword. However there is a huge limiting factor and that is only the main branch and the current commit is searched. To help combat this, the `search_github_code` task retrieves the username of the repositories which were found to contain a unique keyword. Then for each user, it retrieves the repositories belonging to that user and calls `gitleaks` on each repository. This way it's sort of working in a pseudo `gitrob` fashion. 
- `lib/tasks/search_github.rb` -> Removed as this (legacy) task was performing the majority of the same functionality above.
- `lib/workflows/profile_organization_external_light_active.yml` ->  Updated to support the new `gitleaks` task.
- `util/bootstrap.sh` -> Installs the `gitleaks` binary.

Best regards,
Maxim

### CODE REVIEW

**Comment by jcran** [comment]
Very nice description of changes, good work on this @m-q-t.

File: lib/tasks/gather_github_repositories.rb [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+        begin
+            if name.nil?
+              return client.repos
+            end # if no github account is provided, return all repos belonging to access token
+            return nil unless account_exists?(name)
+
+            repositories = if org?(name)
+                             client.org_repos(name, { 'type' => 'all' })
+                           else
+                             client.repos(name)
```
Comment (shpendk) [inline lib/tasks/gather_github_repositories.rb:58]: This does not retrieve private repositories

**Review by shpendk — COMMENTED** [review]
Small change as discussed

**Comment by jcran** [comment]
Summary of changes in https://github.com/intrigueio/intrigue-core/pull/364/commits/11f4b93621b76435d1b7dfaa986e14af294d556c

 - remove auto_paginate from oktokit (was getting a lot of abuse messages / missing data)
 - build manual pagination, with some built-in weights
 - add github_account_to_repositories using (only) oktokit ... since task will always be authenticated 
 - ensure suspicious_commit issues are created independently
 - ensure suspicious_commit issues are created with a source
 - split up tasks a bit (thinking:  better to use sidekiq's threading with one task / gitleaks-on-repo vs thread in-task ... less brittle)
 - adjust GithubRepository format to ([account]/[repo]) .. given that we have a named entity (Github*) this generally works (discussion needed)
 - various other minor changes to support improved git* workflows 

 Future... 
  - remove gather_github_repositories ? 
  - remove gitrob

File: lib/tasks/gather_github_repositories.rb (hunk 1 above)
Comment (m-q-t) [reply lib/tasks/gather_github_repositories.rb]: fixed

**Pull request merged**